tuner/tuner/libtuner.py

```diff
@@ -749,6 +750,168 @@ def collision_handler(index_hash_list: list[tuple[int, str]]) -> tuple[bool, lis
+    baseline_results = multiprocess_progress_wrapper(
+        num_worker=len(devices),
+        task_list=task_list,
+        function=run_iree_benchmark_module_command,
+        initializer=init_worker_context,
+        initializer_inputs=(worker_context_queue,),
+    )
+    return baseline_results
+
+
+class BaselineResultHandler:
+    def __init__(self) -> None:
+        # Maps device IDs to a list of `BenchmarkResult`.
+        self.device_baseline_results: dict[str, list[BenchmarkResult]] = defaultdict(
+            list
+        )
+
+    def add_run(self, results: list[BenchmarkResult]) -> None:
+        for result in results:
+            self.device_baseline_results[result.device_id].append(result)
+
+    def are_baseline_devices_unique(self, results: list[BenchmarkResult]) -> bool:
+        return len(results) == len(set(result.device_id for result in results))
+
+    def get_valid_time_ms(self, device_id: str) -> list[float]:
+        return [
+            result.time
+            for result in self.device_baseline_results.get(device_id, [])
+            if math.isfinite(result.time)
+        ]
+
+    def num_successful_runs(self, device_id: str) -> int:
+        return len(self.get_valid_time_ms(device_id))
+
+    def get_average_result_ms(self, device_id: str) -> Optional[float]:
+        valid_times = self.get_valid_time_ms(device_id)
+        if valid_times:
+            return sum(valid_times) / len(valid_times)
+        return None
+
+    def detect_regressions(
+        self,
+        baseline_results: list[BenchmarkResult],
+        threshold: float = 1.03,
+    ) -> list[str]:
+        regressions = []
+        for result in baseline_results:
+            if not math.isfinite(result.time):
+                continue
+
+            baseline_avg = self.get_average_result_ms(result.device_id)
+            if baseline_avg is not None and result.time > baseline_avg * threshold:
+                regressions.append(result.device_id)
+                logging.warning(
+                    f"Performance regression detected on device {result.device_id}: "
+                    f"Stored average baseline time = {baseline_avg:.2f} ms, "
+                    f"New baseline time = {result.time:.2f} ms, "
+                    f"Slower by {((result.time - baseline_avg) / baseline_avg) * 100:.2f}%"
+                )
+
+        return regressions
+
+    def is_valid(self) -> bool:
+        """
+        Check if there are any valid finite baseline time recorded.
+        Return True if at least a valid (finite) baseline time recorded,
+        otherwise False.
+        This method determines whether the baseline data is available for computations
+        such as calculating speedup.
+        """
+        return any(
+            self.get_valid_time_ms(device_id)
+            for device_id in self.device_baseline_results
+        )
+
+    def is_valid_for_device(self, device_id: str) -> bool:
+        return bool(self.get_valid_time_ms(device_id))
```

Comment:
Maybe we can remove this function and do `num_successful_runs(device_id) != 0` instead?